avatica/src/main/java/org/apache/calcite/avatica/AvaticaUtils.java

```diff
@@ -200,25 +202,32 @@ public static Class box(Class clazz) {
 
   /** Reads the contents of an input stream and returns as a string. */
   public static String readFully(InputStream inputStream) throws IOException {
-    return _readFully(inputStream).toString();
+    final byte[] bytes = _readFully(inputStream);
+    return new String(bytes, 0, bytes.length, StandardCharsets.UTF_8);
   }
 
   public static byte[] readFullyToBytes(InputStream inputStream) throws IOException {
-    return _readFully(inputStream).toByteArray();
+    return _readFully(inputStream);
   }
 
-  /** Reads the contents of an input stream and returns a ByteArrayOutputStrema. */
-  static ByteArrayOutputStream _readFully(InputStream inputStream) throws IOException {
+  /**
+   * Reads the contents of an input stream and returns a byte array.
+   *
+   * @param inputStream the input to read from.
+   * @return A byte array whose length is equal to the number of bytes contained.
+   */
+  static byte[] _readFully(InputStream inputStream) throws IOException {
     final byte[] bytes = new byte[4096];
-    final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    // Avoid the synchronization on ByteArrayOutputStream
+    final UnsynchronizedBuffer.Writer writer = new UnsynchronizedBuffer.Writer(4096);
```

Comment:
It would be good if you could actually benchmark that.
There's another variation: instead of trying to maintain a single `byte[]` result, one might collect `List<byte[]>`, then merge them back at the final `toByteArray` step.

Have you considered that?